### PR TITLE
Basic UDF support for Node API

### PR DESCRIPTION
**Pull request by hannes**
This PR adds basic JS UDFs for the Node API. Here is a minimal example:

```JavaScript
db = new duckdb.Database(':memory:');
db.register("my_udf", "integer", (x) => x+1);
db.all("select my_udf(42) v", function(err, rows) {
   console.log(rows[0]);
});
db.unregister("my_udf");
```

### CODE REVIEW

File: tools/nodejs/src/database.cpp [hunk 1]
```diff
@@ -28,9 +28,11 @@ struct OpenTask : public Task {
 
 	void DoWork() override {
 		try {
-			Get<Database>().database = duckdb::make_unique<duckdb::DuckDB>(filename);
-			duckdb::ParquetExtension extension;
-			extension.Load(*Get<Database>().database);
+			duckdb::DBConfig config;
+			config.maximum_threads = 1;
+			Get<Database>().database = duckdb::make_unique<duckdb::DuckDB>(filename, &config);
+			//			duckdb::ParquetExtension extension;
```
Comment (samansmink) [inline tools/nodejs/src/database.cpp:34]: Should parquet extension remain disabled here?

File: tools/nodejs/src/connection.cpp [hunk 2]
```diff
@@ -66,6 +70,256 @@ Napi::Value Connection::Prepare(const Napi::CallbackInfo &info) {
...
+	if (info.Length() > 3 && info[3].IsFunction()) {
+		completion_callback = info[3].As<Napi::Function>();
+	}
+
+	if (udfs.find(name) != udfs.end()) {
+		Napi::TypeError::New(env, "UDF with this name already exists").ThrowAsJavaScriptException();
+		return env.Null();
+	}
+
+	auto udf = DuckDBNodeUDFFUnction::New(env, udf_callback, "duckdb_node_udf" + name, 0, 1, nullptr,
```
Comment (samansmink) [inline tools/nodejs/src/connection.cpp:266]: I think this 1 is connected to the thread limit of 1 in tools/nodejs/src/database.cpp:32?
Maybe we should make a constexpr of the thread limit and use it in both places.

**Review by samansmink — COMMENTED** [review]
My node API experience is very limited but I added two small comments. The rest looks good to me!

File: tools/nodejs/src/database.cpp (hunk 1 above)
Comment (hannesmuehleisen) [reply tools/nodejs/src/database.cpp:34]: Good catch that was accidental

**Comment by hannesmuehleisen** [comment]
I seem to have broken something wrt prepare(). Will have a look tomorrow.

File: tools/nodejs/lib/duckdb.js [hunk 3]
```diff
@@ -23,42 +23,130 @@ Connection.prototype.each = function(sql) {
...
+
+            const out_data = ptr_to_arr(data_buffers[descr.ret.data_buffer], descr.ret.physical_type, descr.rows);
+            const out_validity = data_buffers[descr.ret.validity_buffer];
+
+            switch (descr.args.length) {
+                case 0:
+                    for (let i = 0; i < descr.rows; ++i) {
+                        const res = fun();
+                        out_data[i] = res;
+                        out_validity[i] = res == undefined ? 0 : 1;
```
Comment (ankoh) [inline tools/nodejs/lib/duckdb.js:75]: I'd prefer null over undefined here as return type for nulls.
Maybe `res === undefined || res === null` ?

Comment (hannesmuehleisen) [reply tools/nodejs/lib/duckdb.js:75]: So what happens if we get undefined then?

Comment (ankoh) [reply tools/nodejs/lib/duckdb.js:75]: maybe `out_validity[i] = (res == undefined || res === null)? 0 : 1;` ?
I just thought that supporting an actual null as return value might be semantically better.

**Comment by hannesmuehleisen** [comment]
The Node 14 crash seems related to https://github.com/node-ffi-napi/ref-napi/issues/54

**Comment by hannes** [comment]
@Mytherin I think this can be merged?

**Comment by hannes** [comment]
need to fix the accidental checkin of binding.gyp

File: Makefile [hunk 4]
```diff
@@ -206,6 +206,7 @@ format-check-silent:
 	python3 scripts/format.py --all --check --silent
 
 format-fix:
+	rm -rf src/amalgamation/*
```
Comment (Mytherin) [inline Makefile:209]: Should these files not be ignored by the format script already?

**Review by Mytherin — APPROVED** [review]

**Pull request merged**